cloudfiles/interfaces.py

```diff
@@ -150,92 +169,139 @@ def put_file(
-    }
+      return {
+        "Cache-Control": None,
+        "Content-Length": statinfo.st_size,
+        "Content-Type": None,
+        "ETag": None,
+        "Last-Modified": datetime.utcfromtimestamp(statinfo.st_mtime),
+        "Content-Md5": None,
+        "Content-Encoding": encoding,
+        "Content-Disposition": None,
+        "Content-Language": None,
+        "Storage-Class": None,
+        "Request-Charged": None,
+        "Parts-Count": None,
+      }
+
+    if self._lock_dir:
+      lock_path = os.path.join(self._lock_dir,
+                               os.path.abspath(path).replace(os.sep, "=+"))
+      rw_lock = fasteners.InterProcessReaderWriterLock(lock_path)
+      with rw_lock.read_lock():
+        return do_head()
+    else:
+      return do_head()
 
   def get_file(self, file_path, start=None, end=None):
     global EXT_TEST_SEQUENCE
     global read_file
     path = self.get_path_to_file(file_path)
 
-    with EXT_TEST_SEQUENCE_LOCK:
-      seq = list(EXT_TEST_SEQUENCE)
+    def do_get_file():
+      with EXT_TEST_SEQUENCE_LOCK:
+        seq = list(EXT_TEST_SEQUENCE)
 
-    i = 0
-    try:
-      for i, (ext, encoding) in enumerate(seq):
-        try:
-          return read_file(path + ext, encoding, start, end)
-        except FileNotFoundError:
-          continue
-    finally:
-      if i > 0:
-        with EXT_TEST_SEQUENCE_LOCK:
-          EXT_TEST_SEQUENCE.insert(0, EXT_TEST_SEQUENCE.pop(i))
+      i = 0
+      try:
+        for i, (ext, encoding) in enumerate(seq):
+          try:
+            return read_file(path + ext, encoding, start, end)
+          except FileNotFoundError:
+            continue
+      finally:
+        if i > 0:
+          with EXT_TEST_SEQUENCE_LOCK:
+            EXT_TEST_SEQUENCE.insert(0, EXT_TEST_SEQUENCE.pop(i))
 
-    return (None, None, None, None)
+      return (None, None, None, None)
+
+    if self._lock_dir:
+      lock_path = os.path.join(self._lock_dir,
+                               os.path.abspath(path).replace(os.sep, "=+"))
+      rw_lock = fasteners.InterProcessReaderWriterLock(lock_path)
+      with rw_lock.read_lock():
+        return do_get_file()
+    else:
+      return do_get_file()
 
   def size(self, file_path):
     path = self.get_path_to_file(file_path)
 
-    with EXT_TEST_SEQUENCE_LOCK:
-      exts = [ pair[0] for pair in EXT_TEST_SEQUENCE ]
-    errors = (FileNotFoundError,)
+    def do_size():
+      with EXT_TEST_SEQUENCE_LOCK:
+        exts = [ pair[0] for pair in EXT_TEST_SEQUENCE ]
+      errors = (FileNotFoundError,)
 
-    for ext in exts:
-      try:
-        return os.path.getsize(path + ext)
-      except errors:
-        continue
+      for ext in exts:
+        try:
+          return os.path.getsize(path + ext)
+        except errors:
+          continue
 
-    return None
+      return None
+
+    if self._lock_dir:
+      lock_path = os.path.join(self._lock_dir,
+                               os.path.abspath(path).replace(os.sep, "=+"))
```

Comment:
This entire pattern can probably be abstracted into a function that accepts a function as input.